File: libs/python/flow_like/langchain.py
"""LangChain integrations for the Flow-Like API.

Provides :class:`FlowLikeChatModel` (a LangChain ``BaseChatModel``) and
:class:`FlowLikeEmbeddings` (a LangChain ``Embeddings`` implementation)
that delegate to the Flow-Like REST endpoints.
"""

from __future__ import annotations

from typing import Any, Optional

import httpx
from langchain_core.callbacks import CallbackManagerForLLMRun
from langchain_core.embeddings import Embeddings
from langchain_core.language_models.chat_models import BaseChatModel
from langchain_core.messages import AIMessage, BaseMessage
from langchain_core.outputs import ChatGeneration, ChatResult

_ROLE_MAP = {
    "human": "user",
    "ai": "assistant",
    "system": "system",
    "tool": "tool",
}


def _build_headers(token: str) -> dict[str, str]:
    """Build HTTP headers for a Flow-Like API request.

    Args:
        token: A PAT (``pat_…``) or API key (``key_…``).

    Returns:
        Header dict with the appropriate auth entry.
    """
    headers = {"Content-Type": "application/json"}
    if token.startswith("pat_"):
        headers["Authorization"] = token
    else:
        headers["X-API-Key"] = token
    return headers


def _to_api_message(msg: BaseMessage) -> dict[str, str]:
    """Convert a LangChain message to the Flow-Like API format.

    Args:
        msg: A LangChain ``BaseMessage`` instance.

    Returns:
        Dict with ``role`` and ``content`` keys.
    """
    return {
        "role": _ROLE_MAP.get(msg.type, "user"),
        "content": str(msg.content),
    }


class FlowLikeChatModel(BaseChatModel):
    """LangChain chat model backed by the Flow-Like completions API.

    Attributes:
        base_url: Flow-Like API base URL.
        token: Authentication token (PAT or API key).
        bit_id: Identifier of the model bit.
        temperature: Sampling temperature override.
        max_tokens: Maximum tokens to generate.
        top_p: Nucleus-sampling probability mass.
    """

    base_url: str
    token: str
    bit_id: str
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    top_p: Optional[float] = None

    @property
    def _llm_type(self) -> str:
        """Return the LLM type identifier used by LangChain."""
        return "flow-like"

    def _generate(
        self,
        messages: list[BaseMessage],
        stop: list[str] | None = None,
        run_manager: CallbackManagerForLLMRun | None = None,
        **kwargs: Any,
    ) -> ChatResult:
        """Synchronously generate a chat completion.

        Args:
            messages: Conversation history as LangChain messages.
            stop: Optional stop sequences.
            run_manager: LangChain callback manager.
            **kwargs: Extra parameters forwarded to the API.

        Returns:
            A ``ChatResult`` containing the model's response.
        """
        body: dict[str, Any] = {
            "messages": [_to_api_message(m) for m in messages],
            "model": self.bit_id,
        }
        if self.temperature is not None:
            body["temperature"] = self.temperature
        if self.max_tokens is not None:
            body["max_tokens"] = self.max_tokens
        if self.top_p is not None:
            body["top_p"] = self.top_p
        if stop:
            body["stop"] = stop

        resp = httpx.post(
            f"{self.base_url}/api/v1/chat/completions",
            headers=_build_headers(self.token),
            json=body,
            timeout=120,
        )
        resp.raise_for_status()
        data = resp.json()

        text = data["choices"][0]["message"]["content"]
        return ChatResult(generations=[ChatGeneration(message=AIMessage(content=text))])


class FlowLikeEmbeddings(Embeddings):
    """LangChain embeddings wrapper backed by the Flow-Like embeddings API."""

    def __init__(self, *, base_url: str, token: str, bit_id: str) -> None:
        """Initialise the embeddings wrapper.

        Args:
            base_url: Flow-Like API base URL.
            token: Authentication token (PAT or API key).
            bit_id: Identifier of the embeddings bit.
        """
        super().__init__()
        self.base_url = base_url
        self.token = token
        self.bit_id = bit_id

    def _embed(self, texts: list[str], embed_type: str) -> list[list[float]]:
        """Send a synchronous embedding request.

        Args:
            texts: Input texts to embed.
            embed_type: Either ``"document"`` or ``"query"``.

        Returns:
            List of embedding vectors.
        """
        resp = httpx.post(
            f"{self.base_url}/api/v1/embeddings/embed",
            headers=_build_headers(self.token),
            json={"model": self.bit_id, "input": texts, "embed_type": embed_type},
            timeout=120,
        )
        resp.raise_for_status()
        return resp.json()["embeddings"]

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        """Embed a list of documents.

        Args:
            texts: Document texts to embed.

        Returns:
            List of embedding vectors, one per document.
        """
        return self._embed(texts, "document")

    def embed_query(self, text: str) -> list[float]:
        """Embed a single query string.

        Args:
            text: The query text.

        Returns:
            Embedding vector for the query.
        """
        return self._embed([text], "query")[0]

    async def _aembed(self, texts: list[str], embed_type: str) -> list[list[float]]:
        """Send an asynchronous embedding request.

        Args:
            texts: Input texts to embed.
            embed_type: Either ``"document"`` or ``"query"``.

        Returns:
            List of embedding vectors.
        """
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{self.base_url}/api/v1/embeddings/embed",
                headers=_build_headers(self.token),
                json={"model": self.bit_id, "input": texts, "embed_type": embed_type},
                timeout=120,
            )
            resp.raise_for_status()
            return resp.json()["embeddings"]

    async def aembed_documents(self, texts: list[str]) -> list[list[float]]:
        """Asynchronously embed a list of documents.

        Args:
            texts: Document texts to embed.

        Returns:
            List of embedding vectors, one per document.
        """
        return await self._aembed(texts, "document")

    async def aembed_query(self, text: str) -> list[float]:
        """Asynchronously embed a single query string.

        Args:
            text: The query text.

        Returns:
            Embedding vector for the query.
        """
        return (await self._aembed([text], "query"))[0]

File: libs/platform/nodejs/src/langchain.ts
import type { CallbackManagerForLLMRun } from "@langchain/core/callbacks/manager";
import { Embeddings, type EmbeddingsParams } from "@langchain/core/embeddings";
import {
	BaseChatModel,
	type BaseChatModelParams,
} from "@langchain/core/language_models/chat_models";
import { AIMessage, type BaseMessage } from "@langchain/core/messages";
import type { ChatResult } from "@langchain/core/outputs";
import { stripTrailingSlashes } from "./client.js";

export interface FlowLikeChatModelParams extends BaseChatModelParams {
	baseUrl: string;
	token: string;
	bitId: string;
	temperature?: number;
	maxTokens?: number;
	topP?: number;
	stop?: string[];
}

export interface FlowLikeEmbeddingsParams extends EmbeddingsParams {
	baseUrl: string;
	token: string;
	bitId: string;
}

function buildHeaders(token: string): Record<string, string> {
	const headers: Record<string, string> = {
		"Content-Type": "application/json",
	};
	if (token.startsWith("pat_")) {
		headers["Authorization"] = token;
	} else {
		headers["X-API-Key"] = token;
	}
	return headers;
}

const ROLE_MAP: Record<string, string> = {
	human: "user",
	ai: "assistant",
	system: "system",
	tool: "tool",
};

export class FlowLikeChatModel extends BaseChatModel {
	lc_namespace = ["flow-like"];
	private baseUrl: string;
	private token: string;
	private bitId: string;
	private temperature?: number;
	private maxTokens?: number;
	private topP?: number;
	private stopSequences?: string[];

	constructor(params: FlowLikeChatModelParams) {
		super(params);
		this.baseUrl = stripTrailingSlashes(params.baseUrl);
		this.token = params.token;
		this.bitId = params.bitId;
		this.temperature = params.temperature;
		this.maxTokens = params.maxTokens;
		this.topP = params.topP;
		this.stopSequences = params.stop;
	}

	_llmType(): string {
		return "flow-like";
	}

	async _generate(
		messages: BaseMessage[],
		options: this["ParsedCallOptions"],
		_runManager?: CallbackManagerForLLMRun,
	): Promise<ChatResult> {
		const body: Record<string, unknown> = {
			messages: messages.map((m) => ({
				role: ROLE_MAP[m._getType()] ?? "user",
				content:
					typeof m.content === "string" ? m.content : JSON.stringify(m.content),
			})),
			model: this.bitId,
		};
		if (this.temperature != null) body.temperature = this.temperature;
		if (this.maxTokens != null) body.max_tokens = this.maxTokens;
		if (this.topP != null) body.top_p = this.topP;
		if (this.stopSequences) body.stop = this.stopSequences;

		const res = await fetch(`${this.baseUrl}/api/v1/chat/completions`, {
			method: "POST",
			headers: buildHeaders(this.token),
			body: JSON.stringify(body),
			signal: options?.signal as AbortSignal | undefined,
		});

		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`Flow-Like API error ${res.status}: ${text}`);
		}

		const data = (await res.json()) as Record<string, any>;
		const content = data.choices?.[0]?.message?.content ?? "";

		return {
			generations: [{ text: content, message: new AIMessage(content) }],
		};
	}
}

export class FlowLikeEmbeddings extends Embeddings {
	private baseUrl: string;
	private token: string;
	private bitId: string;

	constructor(params: FlowLikeEmbeddingsParams) {
		super(params);
		this.baseUrl = stripTrailingSlashes(params.baseUrl);
		this.token = params.token;
		this.bitId = params.bitId;
	}

	private async callApi(
		input: string[],
		embedType: "query" | "document",
	): Promise<number[][]> {
		const res = await fetch(`${this.baseUrl}/api/v1/embeddings/embed`, {
			method: "POST",
			headers: buildHeaders(this.token),
			body: JSON.stringify({
				model: this.bitId,
				input,
				embed_type: embedType,
			}),
		});

		if (!res.ok) {
			const text = await res.text().catch(() => "");
			throw new Error(`Flow-Like API error ${res.status}: ${text}`);
		}

		const data = (await res.json()) as Record<string, any>;
		return data.embeddings;
	}

	async embedDocuments(documents: string[]): Promise<number[][]> {
		return this.callApi(documents, "document");
	}

	async embedQuery(document: string): Promise<number[]> {
		const result = await this.callApi([document], "query");
		return result[0];
	}
}
